lib/maths/CAdaptiveBucketing.cc

```diff
@@ -525,22 +627,94 @@ bool CAdaptiveBucketing::bucket(core_t::TTime time, std::size_t& result) const {
                   << m_Endpoints[n - 1] << ")");
         return false;
     }
-
     result = i - 1;
     return true;
 }
 
 uint64_t CAdaptiveBucketing::checksum(uint64_t seed) const {
     seed = CChecksum::calculate(seed, m_DecayRate);
     seed = CChecksum::calculate(seed, m_MinimumBucketLength);
+    seed = CChecksum::calculate(seed, m_TargetSize);
+    seed = CChecksum::calculate(seed, m_LastLargeErrorBucket);
+    seed = CChecksum::calculate(seed, m_LastLargeErrorPeriod);
+    seed = CChecksum::calculate(
+        seed, m_LargeErrorCountSignificances.toDelimited(significanceToDelimited));
+    seed = CChecksum::calculate(seed, m_MeanWeight);
     seed = CChecksum::calculate(seed, m_Endpoints);
-    return CChecksum::calculate(seed, m_Centres);
+    seed = CChecksum::calculate(seed, m_Centres);
+    seed = CChecksum::calculate(seed, m_LargeErrorCounts);
+    seed = CChecksum::calculate(seed, m_MeanDesiredDisplacement);
+    return CChecksum::calculate(seed, m_MeanAbsDesiredDisplacement);
 }
 
 std::size_t CAdaptiveBucketing::memoryUsage() const {
     std::size_t mem{core::CMemory::dynamicSize(m_Endpoints)};
     mem += core::CMemory::dynamicSize(m_Centres);
+    mem += core::CMemory::dynamicSize(m_LargeErrorCounts);
     return mem;
 }
+
+void CAdaptiveBucketing::maybeSplitBucket() {
+    double largeErrorCount{std::accumulate(m_LargeErrorCounts.begin(),
+                                           m_LargeErrorCounts.end(), 0.0)};
+    if (largeErrorCount >= MINIMUM_LARGE_ERROR_COUNT_TO_SPLIT) {
+        m_LargeErrorCountSignificances = TFloatUInt32PrMinAccumulator{};
+
+        // We compute the right tail p-value of the count of large errors
+        // in a bucket for the null hypothesis that they are uniformly
+        // distributed on the total bucketed period and split if this is
+        // less than a specified threshold.
+        double period{m_Endpoints[m_Endpoints.size() - 1] - m_Endpoints[0]};
+        for (std::size_t i = 1u; i < m_Endpoints.size(); ++i) {
+            double interval{m_Endpoints[i] - m_Endpoints[i - 1]};
+            try {
+                boost::math::binomial binomial{largeErrorCount, interval / period};
+                double oneMinusCdf{
+                    CTools::safeCdfComplement(binomial, m_LargeErrorCounts[i - 1])};
+                m_LargeErrorCountSignificances.add({oneMinusCdf, i - 1});
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Failed to calculate splitting significance: " << e.what());
```

Comment:
in case we hit this, does it make sense to continue the for-loop?

Reply:
It's somewhat moot, i.e. this really shouldn't happen, but I think it is ok to continue: we'll find the highest significance for which there was no error.